### PR TITLE
659.2 sync logger error code

### LINKED ISSUE #659: Compilation for sync settings
This branch does not compile: https://github.com/openmsupply/open-msupply/pulls

Need to propagate mutation/query changes to syncSettings.

Also as part of this would be great to do error mapping in structured way.

Ended up with more changes then I wanted to add to one PR, so broken down into:

- [ ] https://github.com/openmsupply/open-msupply/pull/671
   * Want to include url and flatten down MappedError, move variants of SyncErrorV5 to the same level as MappedError.  
   * Reformat thiserror derives slightly, they don't need to always provide 'source' as this can be traversed (like what anyhow::Error does)
   * Add specific error for SyncError rather then anyhow::Error (easier to map underlying SynApiError and better formatting for thiserror
- [ ] https://github.com/openmsupply/open-msupply/pull/672
- [ ] https://github.com/openmsupply/open-msupply/pull/673
- [ ] https://github.com/openmsupply/open-msupply/pull/677
- [ ] https://github.com/openmsupply/open-msupply/pull/679

### PR DESCRIPTION
**Pull request by andreievg**
closes #659 (this is to link the issue, there is a series of PR, as described in the issue)

[HEAD PR LINK](https://github.com/openmsupply/open-msupply/pull/679)
big diff due to one file being re-formatted, i made a comment

### Changes

Added sync api error codes, so that we can record error in sync log and then return them in strong typed way in API

Added test for sync status (and noticed they are not passing for initialisation status, due to postgres sorting, added `apply_sort_asc_nulls_first` to fix it)

Added documentation of how logger logs and sync status deduces statuses

Mapping from SyncError to sync log code

### CODE REVIEW

File: server/service/src/sync/sync_status/test.rs [hunk 1]
```diff
@@ -1,13 +1,16 @@
 use std::{collections::HashMap, sync::Arc};
```
Comment (andreievg) [inline server/service/src/sync/sync_status/test.rs:1]: Changes in this file are pretty much all format, i had a very long string previously and cargo fmt fails to format file if there is a long string inside chaining : /.

File: server/service/src/sync/sync_status/logger.rs [hunk 2]
```diff
@@ -209,9 +220,210 @@ impl<'a> SyncLogger<'a> {
...
+            | SyncError::RemotePushError(RemotePushError::SyncApiError(error))
+            | SyncError::WaitForIntegrationError(WaitForIntegrationError::SyncApiError(error)) => {
+                error
+            }
+            // Integration timeout reached
+            SyncError::WaitForIntegrationError(_) => {
+                return Self::new(SyncLogRowErrorCode::IntegrationTimeoutReached, sync_error)
+            }
+            // Internal errors
+            _ => return Self::message_only(sync_error),
```
Comment (andreievg) [inline server/service/src/sync/sync_status/logger.rs:245]: Still a chance to miss an error here, similar to error downcasting, but this syntax and logic I think is clearer

**Review by mark-prins — APPROVED** [review]
👍  LGTM!
I may have glossed over the last few tests.. the general approach looks good to me

**Comment by andreievg** [comment]
closing, merged with this: https://github.com/openmsupply/open-msupply/pull/679

**Pull request closed without merging**